sdk/src/main/java/com/hedera/hashgraph/sdk/Transaction.java

```diff
@@ -803,16 +805,73 @@ public T freezeWith(@Nullable Client client) {
         return (T) this;
     }
 
-    void buildTransactions(int untilIndex) {
-        for (var i = transactions.size(); i < untilIndex; ++i) {
-            transactions.add(com.hedera.hashgraph.sdk.proto.Transaction.newBuilder()
-                .setSignedTransactionBytes(
-                    signedTransactions.get(i)
-                        .setSigMap(signatures.get(i))
-                        .build()
-                        .toByteString()
-                ).build());
+    void buildAllTransactions() {
+        for (var i = 0; i < signedTransactions.size(); ++i) {
+            buildTransaction(i);
+        }
+    }
+
+    /**
+     * Will build the specific transaction at {@code index} and will fill with `null` for any empty indices before it
+     *
+     * @param index
+     */
+    void buildTransaction(int index) {
+        if (transactions.size() < index) {
+            for (var i = transactions.size(); i < index - 1; ++i) {
+                transactions.add(null);
+            }
+        } else if (
+            transactions.size() > index &&
+                transactions.get(index) != null &&
+                !transactions.get(index).getSignedTransactionBytes().isEmpty()
+        ) {
+            return;
+        }
+
+        signTransaction(index);
+
+        transactions.add(com.hedera.hashgraph.sdk.proto.Transaction.newBuilder()
+            .setSignedTransactionBytes(
+                signedTransactions.get(index)
+                    .setSigMap(signatures.get(index))
+                    .build()
+                    .toByteString()
+            ).build());
+    }
+
+    /**
+     * Will sign the specific transaction at {@code index} and will fill with `null` for any empty indices before it
+     *
+     * @param index
+     */
+    void signTransaction(int index) {
+        if (signatures.get(index).getSigPairCount() != 0) {
```

Comment:
It's possible I'm being dumb and missing something obvious, but when do the null signatures get added to the `signatures` list?  It looks like the very first thing this function does is `signatures.get(index)`, but the doc comment seems to imply that it should be adding nulls to the list from `signatures.size()` up to `index` first.